pkg/storage/detected/fields.go

```diff
@@ -0,0 +1,42 @@
+package detected
+
+import (
+	"github.com/grafana/loki/v3/pkg/logproto"
+)
+
+func MergeFields(fields []*logproto.DetectedField, limit uint32) []*logproto.DetectedField {
+	mergedFields := make(map[string]*logproto.DetectedField, limit)
+	foundFields := uint32(0)
+
+	for _, field := range fields {
+		if field == nil {
+			continue
+		}
+
+		//TODO(twhitney): this will take the first N up to limit, is there a better
+		//way to rank the fields to make sure we get the most interesting ones?
+		f, ok := mergedFields[field.Label]
+		if !ok && foundFields < limit {
+			mergedFields[field.Label] = field
+			foundFields++
+			continue
+		}
+
+		//seeing the same field again, update the cardinality if it's higher
+		//this is an estimation, as the true cardinality could be greater
+		//than either of the seen values, but will never be less
+		if ok {
+			curCard, newCard := f.Cardinality, field.Cardinality
```

Comment:
if you merge multiple timerange which isn't the cardinality additive ?

Reply:
because you don't know if you've seen different values in each timerange, for example

timerange 1 (cardinality 3)
```
flavor=sweet
flavor=sour
flavor=sweet
flavor=spicy
```

timerange 2 (cardinality 2)
```
flavor=sweet
flavor=sweet
flavor=bland
```

the safest bet is to take the highest cardinality seen (though, like in this example, the true cardinality is actually 4). since we're discarding the actual values, there's no way to be 100% accurate here

Reply:
Sounds like you might want to run the sketch in the frontend may be. Happy to start like this though.

But if you can stream 100k logs back to frontend and do sketch there then it will be more accurate. may be querier could do the parsing of fields to absord most of the CPU work.

Reply:
Cyril suggest async looking in to some ways to merge the sketches in the frontend, so I'm going to investigate that a bit.

Reply:
I've changed this code to merge sketches and get more accurate cardinality estimations